## CMPINF 2100: Homework 05

Vishruth Reddy
## Assigned: Tuesday of Week 05 at 11:00PM

## DUE: Tuesday of Week 06 at 11:59PM

## Overview

This assignment is more open ended than the previous ones. You must explore three data sets using the skills you have learned up to this point in the course. The Week 05 lecture material discusses the ESSENTIAL Pandas methods methods that you must ALWAYS use when BEGINNING to explore data. You must use those methods here in this assignment. You must also use SPLIT-APPLY-COMBINE to explore if summary statistics change across grouping variables. The last problem requires you to use the ESSENTIAL Pandas methods to begin exploring data associated with one of the final projects of your choosing.

**You must download the 3 data sets provided in the Canvas assignment page and save them to the appropriate directory on your computer.**

## Required tasks for Problem 01 through 03

For each of the 3 assigned data sets you must perform the following ESSENTIAL activities:

* Display the number of rows and columns in the dataset  
* Display the names of the columns and their associated data types  
* Display the number of unique values for each column  
* Display the number of MISSING values for each column  
* Display summary statistics appropriate for the data type for each column
* Count the number of rows associated with each unique value for:
    * Non-numeric columns  
    * Numeric columns with low number of unique values  

You must also execute SPLIT-APPLY-COMBINE to summarize the data based on grouping variables. Specifically, you must group by non-numeric columns and calculate the following summary statistics:
* Calculate the total number of rows for numeric columns  
* Calculate the number of non missing rows for numeric columns
* Calculate the mean and standard deviation of numeric columns
* Calculate the standard error on the mean of numeric columns
* Calculate the number of unique values for non-numeric columns

You should consider grouping by numeric columns that have low number of unique values.  

For each grouped DataFrame, you must answer the following:  
* Does the average of the numeric variables CHANGE across the groups?  

### Collaborators

None
## Problem 00

You will work with the NumPy and Pandas modules in this assignment.

Import NumPy and Pandas using their commonly accepted aliases.

#### 00) - SOLUTION

In [1]:
import numpy as np
import pandas as pd

## Problem 01

Download `hw05_prob_01.csv` from the Canvas site. Save the CSV file in the same directory as this Jupyter notebook.

Read in the `hw05_prob_01.csv` CSV file and assign it to the `df01` object.

Explore the `df01` object following **ALL** required actions listed previously.

You must add more code cells and markdown cells to complete this problem.

#### 01) - SOLUTION

In [2]:
df01 = pd.read_csv('hw05_prob_01.csv')
df01

A   B   C         x
0    a1  b0  C1 -6.240387
1    a1  b0  C1 -4.860661
2    a1  b0  C1 -5.066712
3    a1  b0  C1 -5.635156
4    a1  b0  C1 -4.893876
..   ..  ..  ..       ...
125  a2  b1  C2  1.155377
126  a2  b1  C2  2.818466
127  a2  b1  C2  3.294759
128  a2  b1  C2  2.398190
129  a2  b1  C2  2.527589

[130 rows x 4 columns]

In [3]:
df01.shape

(130, 4)

No. of Rows = 130
No. of Columns = 4

In [4]:
df01.dtypes

A     object
B     object
C     object
x    float64
dtype: object

The column names are `A, B, C and x` and their associated data types are `object, object, object and float` respectively.

In [5]:
df01.nunique()

A      2
B      2
C      2
x    130
dtype: int64

The number of unique values with the exclusion of `null` values for `A, B, C, and x` are `2, 2, 2, and 130` respectively.

In [6]:
df01.nunique(dropna=False)

A      2
B      2
C      2
x    130
dtype: int64

The number of unique values remains the same with the specification of the `dropna` parameter.

In [7]:
df01.isna().sum()

A    0
B    0
C    0
x    0
dtype: int64

There are no missing values in any column.

In [8]:
df01.describe()

x
count  130.000000
mean     2.155995
std      6.015095
min     -7.185043
25%     -0.715064
50%      0.828413
75%      4.375242
max     15.382198

The above cell shows the summary statistics of column `x` only as it is the only column with `numeric (here, float datatype)` values.

In [9]:
df01.describe(include='object')

A    B    C
count   130  130  130
unique    2    2    2
top      a1   b1   C2
freq     68   78   72

The above cell shows the summary statitics of columns `A, B, C` only as they are the only columns of `object datatype`. The only difference in the previous 2 cells is that we passed a parameter `include='object'`. This gives different summary stats as the columns don't contain numeric values and this kind of summary makes more sense than just providing `NaN` as shown below.

In [10]:
df01.describe(include='all')

A    B    C           x
count   130  130  130  130.000000
unique    2    2    2         NaN
top      a1   b1   C2         NaN
freq     68   78   72         NaN
mean    NaN  NaN  NaN    2.155995
std     NaN  NaN  NaN    6.015095
min     NaN  NaN  NaN   -7.185043
25%     NaN  NaN  NaN   -0.715064
50%     NaN  NaN  NaN    0.828413
75%     NaN  NaN  NaN    4.375242
max     NaN  NaN  NaN   15.382198

In [11]:
df01.A.value_counts()

a1    68
a2    62
Name: A, dtype: int64

In [12]:
df01.B.value_counts()

b1    78
b0    52
Name: B, dtype: int64

In [13]:
df01.C.value_counts()

C2    72
C1    58
Name: C, dtype: int64

In [14]:
df01.x.value_counts()[df01.x.value_counts() < 2]

-6.240387    1
 0.984923    1
 1.190530    1
 0.361055    1
 1.474326    1
            ..
-0.512551    1
-1.259009    1
 0.461444    1
 0.156642    1
 2.527589    1
Name: x, Length: 130, dtype: int64

In the above cell, I set the unique count threshold to be `2`. In this example, removing `[df01.x.value_counts() < 2]` wouldn't make a difference as every value is unique.

In [15]:
df01.groupby(['A', 'B', 'C']).\
aggregate(x_rows = ('x', 'size'),
          x_nonmissing = ('x', 'count'),
         x_mean = ('x', 'mean'),
         x_std = ('x', 'std'),
         x_sem = ('x', 'sem'),
         x_unique = ('x', 'nunique'),
         ).reset_index()

A   B   C  x_rows  x_nonmissing     x_mean     x_std     x_sem  x_unique
0  a1  b0  C1       7             7  -5.497979  0.696923  0.263412         7
1  a1  b0  C2      22            22  13.542754  0.872442  0.186005        22
2  a1  b1  C1      25            25  -0.183137  0.654817  0.130963        25
3  a1  b1  C2      14            14   0.818355  0.687713  0.183799        14
4  a2  b0  C1      11            11   5.459841  0.571378  0.172277        11
5  a2  b0  C2      12            12  -0.580973  0.382814  0.110509        12
6  a2  b1  C1      15            15  -5.899792  0.632946  0.163426        15
7  a2  b1  C2      24            24   2.056517  0.767123  0.156588        24

In [16]:
df01.groupby(['x']).\
aggregate(x_rows = ('x', 'size'),
          x_nonmissing = ('x', 'count'),
         x_mean = ('x', 'mean'),
         x_std = ('x', 'std'),
         x_sem = ('x', 'sem'),
         x_unique = ('x', 'nunique'),
         ).reset_index()

x  x_rows  x_nonmissing     x_mean  x_std  x_sem  x_unique
0    -7.185043       1             1  -7.185043    NaN    NaN         1
1    -6.745769       1             1  -6.745769    NaN    NaN         1
2    -6.626860       1             1  -6.626860    NaN    NaN         1
3    -6.624032       1             1  -6.624032    NaN    NaN         1
4    -6.420533       1             1  -6.420533    NaN    NaN         1
..         ...     ...           ...        ...    ...    ...       ...
125  14.273278       1             1  14.273278    NaN    NaN         1
126  14.604840       1             1  14.604840    NaN    NaN         1
127  14.733228       1             1  14.733228    NaN    NaN         1
128  14.977406       1             1  14.977406    NaN    NaN         1
129  15.382198       1             1  15.382198    NaN    NaN         1

[130 rows x 7 columns]

Yes, the mean changes for both the groups.

## Problem 02

Download `hw05_prob_02.csv` from the Canvas site. Save the CSV file in the same directory as this Jupyter notebook.

Read in the `hw05_prob_02.csv` CSV file and assign it to the `df02` object.

Explore the `df02` object following **ALL** required actions listed previously.

You must add more code cells and markdown cells to complete this problem.

#### 02) - SOLUTION

In [17]:
df02 = pd.read_csv('hw05_prob_02.csv')
df02

x         y    A   G
0    0.249430  1.226548  2.0   1
1   -1.049282 -1.100737  NaN   1
2   -0.147525 -0.679438  NaN   1
3    1.053127 -0.475298  NaN   1
4    1.242744 -1.060518  2.0   1
..        ...       ...  ...  ..
363  0.357651 -0.608975  NaN  14
364  0.235176  0.415023  NaN  14
365  0.283675 -0.981750  2.0  14
366  0.929657 -0.546980  2.0  14
367 -1.068429  1.795239  2.0  14

[368 rows x 4 columns]

In [18]:
df02.shape

(368, 4)

No. of Rows = 368; 
No. of Columns = 4

In [19]:
df02.dtypes

x    float64
y    float64
A    float64
G      int64
dtype: object

The column names are `x, y, A, and G` and their associated data types are `float, float, float, and integer` respectively.

In [20]:
df02.nunique()

x    368
y    352
A      2
G     14
dtype: int64

The number of unique values with the exclusion of `null` values for `x, y, A, and G` are `368, 352, 2, and 14` respectively.

In [21]:
df02.nunique(dropna=False)

x    368
y    353
A      3
G     14
dtype: int64

The number of unique values changed to `353 and 3` for `y and A` respectively with the inclusion of null values.

In [22]:
df02.isna().sum()

x     0
y    16
A    21
G     0
dtype: int64

We can see that the number of missing values for `y and A` are `16 and 21` respectively whereas, `x and G` have `NO MISSING VALUES`.

In [23]:
df02.describe()

x           y           A           G
count  368.000000  352.000000  347.000000  368.000000
mean     0.008874    0.287405    1.608069    6.744565
std      1.017163    1.951770    0.488886    4.121450
min     -2.698471   -7.737557    1.000000    1.000000
25%     -0.696962   -0.599940    1.000000    3.000000
50%     -0.027857    0.188779    2.000000    6.000000
75%      0.607358    1.331244    2.000000   10.000000
max      2.936346    7.875071    2.000000   14.000000

The above cell shows the summary statistics of all the columns as all of them have numeric values, that is, int and float.

Using `df02.describe(include='object')` will throw an error stating, `there are no objects to concatenate`. This is true as there are no columns with object data type.

In [24]:
df02.nunique()

x    368
y    352
A      2
G     14
dtype: int64

For problem 2, grouping by non-numeric values isn't possible as we do not have any column with a non-numeric datatype.

To group by numeric columns with low-unique values, we can consider `A and G`.

In [25]:
df02.groupby(['A', 'G']).\
aggregate(x_rows = ('x', 'size'),
          x_nonmissing = ('x', 'count'),
         x_mean = ('x', 'mean'),
         x_std = ('x', 'std'),
         x_sem = ('x', 'sem'),
         x_unique = ('x', 'nunique'),
         ).reset_index()

A   G  x_rows  x_nonmissing    x_mean     x_std     x_sem  x_unique
0   1.0   4      15            15 -0.406053  1.066180  0.275286        15
1   1.0   6      15            15  0.002473  0.783328  0.202254        15
2   1.0   7      25            25 -0.122311  1.176125  0.235225        25
3   1.0   8      14            14  0.159827  1.211836  0.323877        14
4   1.0   9      19            19  0.118602  1.102451  0.252920        19
5   1.0  11      29            29 -0.002671  0.922678  0.171337        29
6   1.0  12       8             8 -0.402446  0.462847  0.163641         8
7   1.0  13      11            11 -0.159503  0.928590  0.279980        11
8   2.0   1      35            35  0.176436  0.914619  0.154599        35
9   2.0   2      39            39  0.234535  1.065475  0.170613        39
10  2.0   3      35            35 -0.032452  1.063964  0.179843        35
11  2.0   5      45            45  0.016668  1.030815  0.153665        45
12  2.0  10      45            45 -0.305826  1.004784  0.149784        45
13  2.0  14      12            12  0.415851  0.856502  0.247251        12

In [26]:
df02.groupby(['A', 'G']).\
aggregate(y_rows = ('y', 'size'),
          y_nonmissing = ('y', 'count'),
         y_mean = ('y', 'mean'),
         y_std = ('y', 'std'),
         y_sem = ('y', 'sem'),
         y_unique = ('y', 'nunique'),
         ).reset_index()

A   G  y_rows  y_nonmissing    y_mean     y_std     y_sem  y_unique
0   1.0   4      15            15  0.050363  0.423809  0.109427        15
1   1.0   6      15            15  0.338426  1.827366  0.471824        15
2   1.0   7      25            25 -0.421756  1.005160  0.201032        25
3   1.0   8      14            14 -0.166025  2.552264  0.682121        14
4   1.0   9      19            19 -0.148560  3.302999  0.757760        19
5   1.0  11      29            15  0.054328  0.184977  0.047761        15
6   1.0  12       8             6  2.871034  1.611799  0.658014         6
7   1.0  13      11            11  0.314568  1.460797  0.440447        11
8   2.0   1      35            35  0.118565  1.108399  0.187354        35
9   2.0   2      39            39  0.660993  1.553211  0.248713        39
10  2.0   3      35            35  0.821434  1.137806  0.192324        35
11  2.0   5      45            45 -0.367768  3.518408  0.524493        45
12  2.0  10      45            45  1.039626  1.258740  0.187642        45
13  2.0  14      12            12  0.253504  1.099218  0.317317        12

Yes, the mean is different.

## Problem 03

Download `hw05_prob_03.csv` from the Canvas site. Save the CSV file in the same directory as this Jupyter notebook.

Read in the `hw05_prob_03.csv` CSV file and assign it to the `df03` object.

Explore the `df03` object following **ALL** required actions listed previously.

You must add more code cells and markdown cells to complete this problem.

#### 03) - SOLUTION

In [27]:
df03 = pd.read_csv('hw05_prob_03.csv')
df03

obs_id key     value
0          1  x1  0.034632
1          2  x1  0.523494
2          3  x1  1.224766
3          4  x1  1.517422
4          5  x1  2.432589
...      ...  ..       ...
1402     197  x7 -8.145633
1403     198  x7       NaN
1404     199  x7 -8.152787
1405     200  x7 -8.063927
1406     201  x7       NaN

[1407 rows x 3 columns]

In [28]:
df03.shape

(1407, 3)

No. of Rows = 1407; 
No. of Columns = 3

In [29]:
df03.dtypes

obs_id      int64
key        object
value     float64
dtype: object

The column names are `obs_id, key, value` and their associated data types are `integer, object, float` respectively.

In [30]:
df03.nunique()

obs_id     201
key          7
value     1379
dtype: int64

The number of unique values with the exclusion of `null` values for `obs_id, key, value` are `201, 7, 1379` respectively.

In [31]:
df03.nunique(dropna=False)

obs_id     201
key          7
value     1380
dtype: int64

The number of unique values changed to `1380` for `value` with the inclusion of null values.

In [32]:
df03.isna().sum()

obs_id     0
key        0
value     28
dtype: int64

There are `28` null values in the `value` column.

In [33]:
df03.describe()

obs_id        value
count  1407.000000  1379.000000
mean    101.000000    -2.222709
std      58.043614     4.465882
min       1.000000   -25.362015
25%      51.000000    -2.683234
50%     101.000000    -0.798222
75%     151.000000     0.250730
max     201.000000     4.209716

In [34]:
df03.describe(include='object')

key
count   1407
unique     7
top       x1
freq     201

The only difference in the previous 2 cells is that we passed a parameter `include='object'` in the latter. This gives different summary stats as the column doesn't contain numeric values in the latter and this kind of summary makes more sense than just providing `NaN` as shown below.

In [35]:
df03.describe(include='all')

obs_id   key        value
count   1407.000000  1407  1379.000000
unique          NaN     7          NaN
top             NaN    x1          NaN
freq            NaN   201          NaN
mean     101.000000   NaN    -2.222709
std       58.043614   NaN     4.465882
min        1.000000   NaN   -25.362015
25%       51.000000   NaN    -2.683234
50%      101.000000   NaN    -0.798222
75%      151.000000   NaN     0.250730
max      201.000000   NaN     4.209716

In [36]:
df03.obs_id.value_counts()

1      7
139    7
129    7
130    7
131    7
      ..
71     7
72     7
73     7
74     7
201    7
Name: obs_id, Length: 201, dtype: int64

In [37]:
df03.key.value_counts()

x1    201
x2    201
x3    201
x4    201
x5    201
x6    201
x7    201
Name: key, dtype: int64

In [38]:
df03.value.value_counts()

 0.034632    1
 1.335656    1
 1.702205    1
 1.526580    1
 1.605637    1
            ..
 0.960716    1
 0.704115    1
 0.280310    1
 0.368433    1
-8.063927    1
Name: value, Length: 1379, dtype: int64

In [39]:
df03.groupby(['key']).\
aggregate(val_rows = ('value', 'size'),
          val_nonmissing = ('value', 'count'),
         val_mean = ('value', 'mean'),
         val_std = ('value', 'std'),
         val_sem = ('value', 'sem'),
         val_unique = ('value', 'nunique'),
         ).reset_index()


key  val_rows  val_nonmissing   val_mean   val_std   val_sem  val_unique
0  x1       201             201  -0.160787  1.735903  0.122441         201
1  x2       201             201   0.100830  0.095558  0.006740         201
2  x3       201             197  -0.050132  1.544633  0.110051         197
3  x4       201             201 -10.187219  5.079706  0.358295         201
4  x5       201             201   1.061053  0.935690  0.065998         201
5  x6       201             197  -1.463831  0.591750  0.042160         197
6  x7       201             181  -5.085382  3.002404  0.223167         181

In [40]:
df03.groupby(['obs_id']).\
aggregate(val_rows = ('value', 'size'),
          val_nonmissing = ('value', 'count'),
         val_mean = ('value', 'mean'),
         val_std = ('value', 'std'),
         val_sem = ('value', 'sem'),
         val_unique = ('value', 'nunique'),
         ).reset_index()

obs_id  val_rows  val_nonmissing  val_mean   val_std   val_sem  \
0         1         7               6 -0.125003  0.376845  0.153846   
1         2         7               6 -0.425234  1.287517  0.525627   
2         3         7               4  0.083965  0.788068  0.394034   
3         4         7               7 -0.188854  1.153164  0.435855   
4         5         7               7 -0.353155  2.137852  0.808032   
..      ...       ...             ...       ...       ...       ...   
196     197         7               7 -3.235109  7.031101  2.657506   
197     198         7               5 -3.131932  9.239653  4.132098   
198     199         7               7 -3.728535  7.722103  2.918680   
199     200         7               6 -3.600374  7.379866  3.012817   
200     201         7               6 -2.675995  7.792097  3.181110   

     val_unique  
0             6  
1             6  
2             4  
3             7  
4             7  
..          ...  
196           7  
197           5  
198           7  
199           6  
200           6  

[201 rows x 7 columns]

Yes, the average changes

## Problem 04

You must download the data associated with one of the Final Projects from the Canvas site. Save the file(s) in the same directory as this Jupyter notebook.

Read in the data associated with one of the Final Projects. You must perform the ESSENTIAL activities listed previously.

You do **NOT** need to SPLIT-APPLY-COMBINE to explore across groups for this assignment.

#### 04) - SOLUTION

In [41]:
songs_url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-01-21/spotify_songs.csv'
sdf = pd.read_csv( songs_url )
sdf

track_id  \
0      6f807x0ima9a1j3VPbc7VN   
1      0r7CVbZTWZgbTCYdfa2P31   
2      1z1Hg7Vb0AhHDiEmnDE79l   
3      75FpbthrwQmzHlBJLuGdC7   
4      1e8PAfcKUYoKkxPhrHqw4x   
...                       ...   
32828  7bxnKAamR3snQ1VGLuVfC1   
32829  5Aevni09Em4575077nkWHz   
32830  7ImMqPP3Q1yfUHvsdn7wEo   
32831  2m69mhnfQ1Oq6lGtXuYhgX   
32832  29zWqhca3zt5NsckZqDf6c   

                                              track_name      track_artist  \
0      I Don't Care (with Justin Bieber) - Loud Luxur...        Ed Sheeran   
1                        Memories - Dillon Francis Remix          Maroon 5   
2                        All the Time - Don Diablo Remix      Zara Larsson   
3                      Call You Mine - Keanu Silva Remix  The Chainsmokers   
4                Someone You Loved - Future Humans Remix     Lewis Capaldi   
...                                                  ...               ...   
32828               City Of Lights - Official Radio Edit      Lush & Simon   
32829                Closer - Sultan & Ned Shepard Remix    Tegan and Sara   
32830                       Sweet Surrender - Radio Edit       Starkillers   
32831                     Only For You - Maor Levi Remix            Mat Zo   
32832                             Typhoon - Original Mix      Julian Calor   

       track_popularity          track_album_id  \
0                    66  2oCs0DGTsRO98Gh5ZSl2Cx   
1                    67  63rPSO264uRjW1X5E6cWv6   
2                    70  1HoSmj2eLcsrR0vE9gThr4   
3                    60  1nqYsOef1yKKuGOVchbsk6   
4                    69  7m7vv9wlQ4i0LFuJiE2zsQ   
...                 ...                     ...   
32828                42  2azRoBBWEEEYhqV6sb7JrT   
32829                20  6kD6KLxj7s8eCE3ABvAyf5   
32830                14  0ltWNSY9JgxoIZO4VzuCa6   
32831                15  1fGrOkHnHJcStl14zNx8Jy   
32832                27  0X3mUOm6MhxR7PzxG95rAo   

                                        track_album_name  \
0      I Don't Care (with Justin Bieber) [Loud Luxury...   
1                        Memories (Dillon Francis Remix)   
2                        All the Time (Don Diablo Remix)   
3                            Call You Mine - The Remixes   
4                Someone You Loved (Future Humans Remix)   
...                                                  ...   
32828                         City Of Lights (Vocal Mix)   
32829                                     Closer Remixed   
32830                       Sweet Surrender (Radio Edit)   
32831                             Only For You (Remixes)   
32832                                      Typhoon/Storm   

      track_album_release_date    playlist_name             playlist_id  \
0                   2019-06-14        Pop Remix  37i9dQZF1DXcZDD7cfEKhW   
1                   2019-12-13        Pop Remix  37i9dQZF1DXcZDD7cfEKhW   
2                   2019-07-05        Pop Remix  37i9dQZF1DXcZDD7cfEKhW   
3                   2019-07-19        Pop Remix  37i9dQZF1DXcZDD7cfEKhW   
4                   2019-03-05        Pop Remix  37i9dQZF1DXcZDD7cfEKhW   
...                        ...              ...                     ...   
32828               2014-04-28  ♥ EDM LOVE 2020  6jI1gFr6ANFtT8MmTvA2Ux   
32829               2013-03-08  ♥ EDM LOVE 2020  6jI1gFr6ANFtT8MmTvA2Ux   
32830               2014-04-21  ♥ EDM LOVE 2020  6jI1gFr6ANFtT8MmTvA2Ux   
32831               2014-01-01  ♥ EDM LOVE 2020  6jI1gFr6ANFtT8MmTvA2Ux   
32832               2014-03-03  ♥ EDM LOVE 2020  6jI1gFr6ANFtT8MmTvA2Ux   

      playlist_genre  ... key  loudness  mode  speechiness  acousticness  \
0                pop  ...   6    -2.634     1       0.0583      0.102000   
1                pop  ...  11    -4.969     1       0.0373      0.072400   
2                pop  ...   1    -3.432     0       0.0742      0.079400   
3                pop  ...   7    -3.778     1       0.1020      0.028700   
4                pop  ...   1    -4.672     1       0.0359      0.080300

In [42]:
sdf.shape

(32833, 23)

No. of Rows = 32833; 
No. of Columns = 23

In [43]:
sdf.dtypes

track_id                     object
track_name                   object
track_artist                 object
track_popularity              int64
track_album_id               object
track_album_name             object
track_album_release_date     object
playlist_name                object
playlist_id                  object
playlist_genre               object
playlist_subgenre            object
danceability                float64
energy                      float64
key                           int64
loudness                    float64
mode                          int64
speechiness                 float64
acousticness                float64
instrumentalness            float64
liveness                    float64
valence                     float64
tempo                       float64
duration_ms                   int64
dtype: object

In [44]:
sdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32833 entries, 0 to 32832
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   track_id                  32833 non-null  object 
 1   track_name                32828 non-null  object 
 2   track_artist              32828 non-null  object 
 3   track_popularity          32833 non-null  int64  
 4   track_album_id            32833 non-null  object 
 5   track_album_name          32828 non-null  object 
 6   track_album_release_date  32833 non-null  object 
 7   playlist_name             32833 non-null  object 
 8   playlist_id               32833 non-null  object 
 9   playlist_genre            32833 non-null  object 
 10  playlist_subgenre         32833 non-null  object 
 11  danceability              32833 non-null  float64
 12  energy                    32833 non-null  float64
 13  key                       32833 non-null  int64  
 14  loudne

We can see that in the Spotify dataset, there are `9 floating values, 4 integer values and 10 object datatypes`.

In [45]:
sdf.nunique()

track_id                    28356
track_name                  23449
track_artist                10692
track_popularity              101
track_album_id              22545
track_album_name            19743
track_album_release_date     4530
playlist_name                 449
playlist_id                   471
playlist_genre                  6
playlist_subgenre              24
danceability                  822
energy                        952
key                            12
loudness                    10222
mode                            2
speechiness                  1270
acousticness                 3731
instrumentalness             4729
liveness                     1624
valence                      1362
tempo                       17684
duration_ms                 19785
dtype: int64

The unique values are displayed above.

In [46]:
sdf.nunique(dropna=False)

track_id                    28356
track_name                  23450
track_artist                10693
track_popularity              101
track_album_id              22545
track_album_name            19744
track_album_release_date     4530
playlist_name                 449
playlist_id                   471
playlist_genre                  6
playlist_subgenre              24
danceability                  822
energy                        952
key                            12
loudness                    10222
mode                            2
speechiness                  1270
acousticness                 3731
instrumentalness             4729
liveness                     1624
valence                      1362
tempo                       17684
duration_ms                 19785
dtype: int64

In [47]:
print(sdf.nunique() != sdf.nunique(dropna=False))

track_id                    False
track_name                   True
track_artist                 True
track_popularity            False
track_album_id              False
track_album_name             True
track_album_release_date    False
playlist_name               False
playlist_id                 False
playlist_genre              False
playlist_subgenre           False
danceability                False
energy                      False
key                         False
loudness                    False
mode                        False
speechiness                 False
acousticness                False
instrumentalness            False
liveness                    False
valence                     False
tempo                       False
duration_ms                 False
dtype: bool


The `track_name, track_artist, track_album_name` have null values too.

In [48]:
sdf.isna().sum()

track_id                    0
track_name                  5
track_artist                5
track_popularity            0
track_album_id              0
track_album_name            5
track_album_release_date    0
playlist_name               0
playlist_id                 0
playlist_genre              0
playlist_subgenre           0
danceability                0
energy                      0
key                         0
loudness                    0
mode                        0
speechiness                 0
acousticness                0
instrumentalness            0
liveness                    0
valence                     0
tempo                       0
duration_ms                 0
dtype: int64

The above cell is another way of showing the presence of null values in the columns.

In [49]:
sdf.describe()

track_popularity  danceability        energy           key  \
count      32833.000000  32833.000000  32833.000000  32833.000000   
mean          42.477081      0.654850      0.698619      5.374471   
std           24.984074      0.145085      0.180910      3.611657   
min            0.000000      0.000000      0.000175      0.000000   
25%           24.000000      0.563000      0.581000      2.000000   
50%           45.000000      0.672000      0.721000      6.000000   
75%           62.000000      0.761000      0.840000      9.000000   
max          100.000000      0.983000      1.000000     11.000000   

           loudness          mode   speechiness  acousticness  \
count  32833.000000  32833.000000  32833.000000  32833.000000   
mean      -6.719499      0.565711      0.107068      0.175334   
std        2.988436      0.495671      0.101314      0.219633   
min      -46.448000      0.000000      0.000000      0.000000   
25%       -8.171000      0.000000      0.041000      0.015100   
50%       -6.166000      1.000000      0.062500      0.080400   
75%       -4.645000      1.000000      0.132000      0.255000   
max        1.275000      1.000000      0.918000      0.994000   

       instrumentalness      liveness       valence         tempo  \
count      32833.000000  32833.000000  32833.000000  32833.000000   
mean           0.084747      0.190176      0.510561    120.881132   
std            0.224230      0.154317      0.233146     26.903624   
min            0.000000      0.000000      0.000000      0.000000   
25%            0.000000      0.092700      0.331000     99.960000   
50%            0.000016      0.127000      0.512000    121.984000   
75%            0.004830      0.248000      0.693000    133.918000   
max            0.994000      0.996000      0.991000    239.440000   

         duration_ms  
count   32833.000000  
mean   225799.811622  
std     59834.006182  
min      4000.000000  
25%    187819.000000  
50%    216000.000000  
75%    253585.000000  
max    517810.000000

In [50]:
sdf.describe(include='object')

track_id track_name   track_artist  \
count                    32833      32828          32828   
unique                   28356      23449          10692   
top     7BKLCZ1jbUBVqRi2FVlTVw     Poison  Martin Garrix   
freq                        10         22            161   

                track_album_id track_album_name track_album_release_date  \
count                    32833            32828                    32833   
unique                   22545            19743                     4530   
top     5L1xcowSxwzFUSJzvyMp48    Greatest Hits               2020-01-10   
freq                        42              139                      270   

          playlist_name             playlist_id playlist_genre  \
count             32833                   32833          32833   
unique              449                     471              6   
top     Indie Poptimism  4JkkvMpVl4lSioqQjeAL0q            edm   
freq                308                     247           6043   

                playlist_subgenre  
count                       32833  
unique                         24  
top     progressive electro house  
freq                         1809

The above cells show the summary stats for numeric and non-numeric values respectively. This is a better way of displaying as we can get rid of the overall statistics with null values as shown below.

In [51]:
sdf.describe(include='all')

track_id track_name   track_artist  track_popularity  \
count                    32833      32828          32828      32833.000000   
unique                   28356      23449          10692               NaN   
top     7BKLCZ1jbUBVqRi2FVlTVw     Poison  Martin Garrix               NaN   
freq                        10         22            161               NaN   
mean                       NaN        NaN            NaN         42.477081   
std                        NaN        NaN            NaN         24.984074   
min                        NaN        NaN            NaN          0.000000   
25%                        NaN        NaN            NaN         24.000000   
50%                        NaN        NaN            NaN         45.000000   
75%                        NaN        NaN            NaN         62.000000   
max                        NaN        NaN            NaN        100.000000   

                track_album_id track_album_name track_album_release_date  \
count                    32833            32828                    32833   
unique                   22545            19743                     4530   
top     5L1xcowSxwzFUSJzvyMp48    Greatest Hits               2020-01-10   
freq                        42              139                      270   
mean                       NaN              NaN                      NaN   
std                        NaN              NaN                      NaN   
min                        NaN              NaN                      NaN   
25%                        NaN              NaN                      NaN   
50%                        NaN              NaN                      NaN   
75%                        NaN              NaN                      NaN   
max                        NaN              NaN                      NaN   

          playlist_name             playlist_id playlist_genre  ...  \
count             32833                   32833          32833  ...   
unique              449                     471              6  ...   
top     Indie Poptimism  4JkkvMpVl4lSioqQjeAL0q            edm  ...   
freq                308                     247           6043  ...   
mean                NaN                     NaN            NaN  ...   
std                 NaN                     NaN            NaN  ...   
min                 NaN                     NaN            NaN  ...   
25%                 NaN                     NaN            NaN  ...   
50%                 NaN                     NaN            NaN  ...   
75%                 NaN                     NaN            NaN  ...   
max                 NaN                     NaN            NaN  ...   

                 key      loudness          mode   speechiness  acousticness  \
count   32833.000000  32833.000000  32833.000000  32833.000000  32833.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        5.374471     -6.719499      0.565711      0.107068      0.175334   
std         3.611657      2.988436      0.495671      0.101314      0.219633   
min         0.000000    -46.448000      0.000000      0.000000      0.000000   
25%         2.000000     -8.171000      0.000000      0.041000      0.015100   
50%         6.000000     -6.166000      1.000000      0.062500      0.080400   
75%         9.000000     -4.645000      1.000000      0.132000      0.255000   
max        11.000000      1.275000      1.000000      0.918000      0.994000   

        instrumentalness      liveness       valence         tempo  \
count       32833.000000  32833.000000  32833.000000  32833.000000   
unique               NaN           NaN           NaN           NaN   
top                  NaN           NaN           NaN           NaN   
freq                 NaN           NaN           NaN           NaN   
mean    

In [52]:
for i in sdf.select_dtypes(exclude=['number']).columns:
    row_count = sdf[i].value_counts()
    print(i,'\n', row_count)

track_id 
 7BKLCZ1jbUBVqRi2FVlTVw    10
14sOS5L36385FJ3OL8hew4     9
3eekarcy7kvN4yt5ZFzltW     9
2Fxmhks0bxGSBdJ92vM42m     8
2tnVG71enUj33Ic2nFN6kZ     8
                          ..
3AKwyujeGxTQSZNbx9Ka3c     1
3ITvHA9zhZZdBJsOsAUegF     1
5gsW8TMmNVnevjq13h0Nlp     1
4pN4icvWINm9uN3bElfbec     1
29zWqhca3zt5NsckZqDf6c     1
Name: track_id, Length: 28356, dtype: int64
track_name 
 Poison                    22
Breathe                   21
Alive                     20
Forever                   20
Stay                      19
                          ..
Quien Eres Tu              1
Como Nadie                 1
Ya No Estás Aquí           1
Ojala                      1
Typhoon - Original Mix     1
Name: track_name, Length: 23449, dtype: int64
track_artist 
 Martin Garrix       161
Queen               136
The Chainsmokers    123
David Guetta        110
Don Omar            102
                   ... 
Underworld            1
The Witches           1
Tess Parks            1
Mick Harvey      

In [53]:
unique_count = {}
for i in sdf.select_dtypes(include=['number']).columns:
    unique_count[i] = sdf[i].nunique()

row_count = sdf[min(unique_count)].value_counts()
print(i,'\n', row_count)

duration_ms 
 0.102000    80
0.128000    80
0.101000    78
0.114000    74
0.141000    71
            ..
0.000075     1
0.000786     1
0.000086     1
0.000026     1
0.000434     1
Name: acousticness, Length: 3731, dtype: int64
